x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/planning/AssignmentPlan.java

```diff
@@ -0,0 +1,323 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.inference.assignment.planning;
+
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.core.Tuple;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * A plan describing how models should be assigned to nodes.
+ */
+public class AssignmentPlan implements Comparable<AssignmentPlan> {
+
+    public record Model(
+        String id,
+        long memoryBytes,
+        int allocations,
+        int threadsPerAllocation,
+        Map<String, Integer> currentAllocationByNodeId
+    ) {
+
+        int getPreviouslyAssignedAllocations() {
+            return currentAllocationByNodeId.values().stream().mapToInt(Integer::intValue).sum();
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hashCode(id);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            Model that = (Model) o;
+            return Objects.equals(id, that.id);
+        }
+
+        @Override
+        public String toString() {
+            return id
+                + " (mem = "
+                + ByteSizeValue.ofBytes(memoryBytes)
+                + ") (allocations = "
+                + allocations
+                + ") (threads_per_allocation = "
+                + threadsPerAllocation
+                + ") (current_allocations = "
+                + currentAllocationByNodeId
+                + ")";
+        }
+    };
+
+    public record Node(String id, long availableMemoryBytes, int cores) {
+
+        @Override
+        public int hashCode() {
+            return Objects.hashCode(id);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            Node that = (Node) o;
+            return Objects.equals(id, that.id);
+        }
```

Comment:
This is sort of sneaky. You do this because you want to have the `Node` object as a hashtable later in the linear solver. But these objects are not actually equivalent as the memory and cores could be drastically different.

Could we instead use the `id` in the hash tables directly?

The same concern goes for the `Model` class.

If we are stuck on this, then at least these classes should be package private.

Reply:
It turns out I can get rid of those sneaky `equals` indeed. The only place where we need to match models/nodes that can be different is when we merge the assignment plan with its preserved allocations. It used to be different which is why I had done this hack. But later refactoring meant we actually don't need it anymore.